crypto/sm2/sm2_pmeth.c

```diff
@@ -0,0 +1,269 @@
+{
+    SM2_PKEY_CTX *dctx = ctx->data;
+    EC_KEY *ec = ctx->pkey->pkey.ec;
+    ECDSA_SIG* der_sig = NULL;
+    const EVP_MD* md = NULL;
+    const char* user_id = NULL;
+
+    if (!sig) {
+        /* ECDSA and SM2 signatures have the same size */
+        *siglen = ECDSA_size(ec);
+        return 1;
+    } else if (*siglen < (size_t)ECDSA_size(ec)) {
+        ECerr(EC_F_PKEY_SM2_SIGN, SM2_R_BUFFER_TOO_SMALL);
+        return 0;
+    }
+
+    md = (dctx->md) ? dctx->md : EVP_sm3();
+
+    user_id = (dctx->user_id) ? dctx->user_id : SM2_DEFAULT_USERID;
+
+    der_sig = SM2_do_sign(ec, md, user_id, tbs, tbslen);
+
+    if(der_sig == NULL)
+       return 0;
+
+    /* Now ASN.1 encode ... */
+
+    *siglen = i2d_ECDSA_SIG(der_sig, &sig);
+
+    return 1;
+}
+
+static int pkey_sm2_verify(EVP_PKEY_CTX *ctx,
+                          const unsigned char *sig, size_t siglen,
+                          const unsigned char *tbs, size_t tbslen)
+{
+    int ret, type;
+    SM2_PKEY_CTX *dctx = ctx->data;
+    EC_KEY *ec = ctx->pkey->pkey.ec;
+
+    if (dctx->md)
+        type = EVP_MD_type(dctx->md);
+    else
+        type = NID_sm3;
+
+    ret = ECDSA_verify(type, tbs, tbslen, sig, siglen, ec);
+
+    return ret;
+}
+
+static int pkey_sm2_ctrl(EVP_PKEY_CTX *ctx, int type, int p1, void *p2)
+{
+    SM2_PKEY_CTX *dctx = ctx->data;
+    EC_GROUP *group;
+
+    switch (type) {
+    case EVP_PKEY_CTRL_EC_PARAMGEN_CURVE_NID:
+        group = EC_GROUP_new_by_curve_name(p1);
+        if (group == NULL) {
+            SM2err(SM2_F_PKEY_SM2_CTRL, SM2_R_INVALID_CURVE);
+            return 0;
+        }
+        EC_GROUP_free(dctx->gen_group);
+        dctx->gen_group = group;
+        return 1;
+
+    case EVP_PKEY_CTRL_MD:
+        {
+        int md_type = EVP_MD_type((const EVP_MD *)p2);
+        if (md_type != NID_sm3 &&
+            md_type != NID_sha256 &&
+            md_type != NID_sha512_256) {
+            SM2err(SM2_F_PKEY_SM2_CTRL, SM2_R_INVALID_DIGEST_TYPE);
+            return 0;
+        }
+        dctx->md = p2;
+        return 1;
+        }
+
+    case EVP_PKEY_CTRL_GET_MD:
+        *(const EVP_MD **)p2 = dctx->md;
+        return 1;
+
+        /* Default behaviour is OK */
+    case EVP_PKEY_CTRL_DIGESTINIT:
+    case EVP_PKEY_CTRL_PKCS7_SIGN:
+    case EVP_PKEY_CTRL_CMS_SIGN:
+        return 1;
+
+    default:
+        return -2;
+
+    }
+}
+
+static int pkey_sm2_ctrl_str(EVP_PKEY_CTX *ctx,
+                            const char *type, const char *value)
+{
+    if (strcmp(type, "ec_paramgen_curve") == 0) {
+        int nid;
+        nid = EC_curve_nist2nid(value);
+        if (nid == NID_undef)
+            nid = OBJ_sn2nid(value);
+        if (nid == NID_undef)
+            nid = OBJ_ln2nid(value);
+        if (nid == NID_undef) {
+            SM2err(SM2_F_PKEY_SM2_CTRL_STR, SM2_R_INVALID_CURVE);
+            return 0;
+        }
+        return EVP_PKEY_CTX_set_ec_paramgen_curve_nid(ctx, nid);
```

Comment:
I think you need new macros for SM2.  `EVP_PKEY_CTX_set_ec_paramgen_curve_nid` is a macro that is specificaly defined for `EVP_PKEY_EC`, and will fail because this pmeth has the PKEY identity `EVP_PKEY_SM2`...  So you'll need a `EVP_PKEY_CTX_set_sm2_paramgen_curve_nid`, and seeing the rest of the control code, I believe you'll need a `EVP_PKEY_CTX_set_sm2_param_enc` as well.

So it seems that checking this pmeth through for this kind of inconsistency might make your pkey life easier ;-)